### PR TITLE
Add support of workload identity to provisioner for CI

### PR DESCRIPTION
**Pull request by surajssd**
- This PR adds support of using workload identity to the CAA application. Using workload identity one does not need to provide any secrets to be able to talk to the Azure API. This features allows us to enable testing Azure provider in CI.
- This PR also enables the Azure test provisioner to create a federated credential which then can be consumed by the CAA application running as a daemonset.
- This PR has a patch file for Azure which can be used to provide the client id of a managed identity.

### CODE REVIEW

**Comment by surajssd** [comment]
## Testing this PR

### Create a RG

```bash
export AZURE_RESOURCE_GROUP=""
export AZURE_REGION="eastus"
```

### Create an identity

```bash
az group create --name "${AZURE_RESOURCE_GROUP}" \
    --location "${AZURE_REGION}"

export AZURE_SUBSCRIPTION_ID=$(az account show --query id --output tsv)

export USER_ASSIGNED_IDENTITY_NAME="caa-${AZURE_RESOURCE_GROUP}"
az identity create \
    --name "${USER_ASSIGNED_IDENTITY_NAME}" \
    --resource-group "${AZURE_RESOURCE_GROUP}" \
    --location "${AZURE_REGION}" \
    --subscription "${AZURE_SUBSCRIPTION_ID}"

export PRINCIPAL_ID="$(az identity show \
    --name "${USER_ASSIGNED_IDENTITY_NAME}" \
    --resource-group "${AZURE_RESOURCE_GROUP}" \
    --subscription "${AZURE_SUBSCRIPTION_ID}" --query principalId -otsv)"

az role assignment create \
    --role Contributor \
    --assignee-object-id "${PRINCIPAL_ID}" \
    --scope "/subscriptions/${AZURE_SUBSCRIPTION_ID}/resourceGroups/${AZURE_RESOURCE_GROUP}"

export AZURE_CLIENT_ID="$(az identity show \
    --resource-group "${AZURE_RESOURCE_GROUP}" \
    --name "${USER_ASSIGNED_IDENTITY_NAME}" --query 'clientId' -otsv)"
```

### Create the config used by provisioner app

```bash
export CLUSTER_NAME="e2e"

cat <<EOF >/tmp/provision_azure.properties
AZURE_CLIENT_ID="${AZURE_CLIENT_ID}"
AZURE_SUBSCRIPTION_ID="${AZURE_SUBSCRIPTION_ID}"
RESOURCE_GROUP_NAME="${AZURE_RESOURCE_GROUP}"
CLUSTER_NAME="${CLUSTER_NAME}"
LOCATION="${AZURE_REGION}"
SSH_KEY_ID="id_rsa.pub"
AZURE_IMAGE_ID="test"
SSH_USERNAME="azureuser"

AZURE_CLI_AUTH="true"
IS_CI_MANAGED_CLUSTER="true"
MANAGED_IDENTITY_NAME="${USER_ASSIGNED_IDENTITY_NAME}"
EOF
```

### Build CAA image

You may have to build the CAA image from the code in this PR for this to work.

First apply this diff (make changes to the image name as necessary):

```diff
diff --git test/provisioner/provision_azure.go test/provisioner/provision_azure.go
index 2da4765..8f9bad9 100644
--- test/provisioner/provision_azure.go
+++ test/provisioner/provision_azure.go
@@ -450,6 +450,12 @@ func (lio *AzureInstallOverlay) Delete(ctx context.Context, cfg *envconf.Config)

 func (lio *AzureInstallOverlay) Edit(ctx context.Context, cfg *envconf.Config, properties map[string]string) error {
        var err error
+
+       // TODO: This is temporary, remove before committing.
+       if err = lio.overlay.SetKustomizeImage("cloud-api-adaptor", "newName", "quay.io/surajd/cloud-api-adaptor"); err != nil {
+               return err
+       }
+
        for k, v := range properties {
                // configMapGenerator
                if isAzureKustomizeConfigMapKey(k) {
```

```bash
export registry=quay.io/surajd
export CLOUD_PROVIDER=azure
make image
```

### Provision the cluster

```bash
cd test/tools
export TEST_PROVISION_FILE=/tmp/provision_azure.properties
export CLOUD_PROVIDER=azure
export BUILTIN_CLOUD_PROVIDERS=azure

make caa-provisioner-cli
./caa-provisioner-cli -action=provision
```

File: entrypoint.sh [hunk 1]
```diff
@@ -52,7 +52,7 @@ aws() {
 }
 
 azure() {
-    test_vars AZURE_CLIENT_ID AZURE_CLIENT_SECRET AZURE_TENANT_ID AZURE_SUBSCRIPTION_ID AZURE_RESOURCE_GROUP AZURE_SUBNET_ID AZURE_IMAGE_ID
+    test_vars AZURE_CLIENT_ID AZURE_TENANT_ID AZURE_SUBSCRIPTION_ID AZURE_RESOURCE_GROUP AZURE_SUBNET_ID AZURE_IMAGE_ID
```
Comment (snir911) [inline entrypoint.sh:55]: tenant_id is needed?
also consider using one_of instead of test_vars , IIUC it suits here (otherwise we can actually remove it)

**Review by snir911 — COMMENTED** [review]
thanks! my understanding here is limited however LGTM overall, add minor comments, also i'd suggest to add your instructions above to the azure docs as part of this PR

File: entrypoint.sh (hunk 1 above)
Comment (surajssd) [reply entrypoint.sh:55]: Yes tenant_id is needed and in service principal case user populates it, in workload identity case it is automatically populated so yes it is needed.

Comment (surajssd) [reply entrypoint.sh:55]: IIUC one_of is for when it is an either / or condition right? We need all the others to be set not an either or condition. But I can think of situations when we need one_of especially with the networking stuff.

File: install/overlays/azure/workload-identity.yaml [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+---
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: cloud-api-adaptor
+  namespace: confidential-containers-system
+  labels:
+    azure.workload.identity/use: "true"
+  annotations:
+    azure.workload.identity/client-id: "00000000-0000-0000-0000-000000000000"
```
[the remark this answers is not in the document]
Comment (surajssd) [reply install/overlays/azure/workload-identity.yaml:20]: Nope, not yet. We need to document this alternate way of providing identity when deploying CAA on AKS.

For a DIY cluster on Azure it a convoluted process.

**Review by kartikjoshi21 — APPROVED** [review]
LGTM

File: entrypoint.sh (hunk 1 above)
Comment (snir911) [reply entrypoint.sh:55]: in the workload identity case AZURE_TENANT_ID is automatically populated as environment variable?

**Comment by bpradipt** [comment]
@snir911 anything pending on this one ?

File: entrypoint.sh (hunk 1 above)
Comment (surajssd) [reply entrypoint.sh:55]: Yep it is automatically populated.

**Pull request merged**